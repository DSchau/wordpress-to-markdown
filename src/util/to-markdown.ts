import * as Turndown from 'turndown';
import { gfm as githubFlavoredMarkdown } from 'turndown-plugin-gfm';

const service = new Turndown({
  bulletListMarker: '-',
  codeBlockStyle: 'fenced',
  headingStyle: 'atx',
  hr: '___'
});

// TODO: it seems like code blocks may be unescaped improperly
service.addRule('code-fencing', {
  filter: ['pre'],
  replacement(content, node) {
    const trimmed = content ? content.trim() : '';
    if (trimmed.match('\n')) {
      const lang = node.getAttribute('lang');
      return ['', '```' + (lang === null ? '' : lang), trimmed, '```', '']
        .join('\n');
    }
    return ['`', trimmed, '\`'].join('');
  }
});

service.use(githubFlavoredMarkdown);

/*
 * This is a pretty naive approach of detecting markdown
 * It works by splitting and if any line contains a markdown header (#)
 * code fenced block, or seperator
 */
export function isMarkdown(html: string): boolean {
  return html.split('\n').some(line => {
    return !!line.match(/(#{1,6}\s?\w+)|```|___/);
  });
}

export function toMarkdown(html: string): string {
  if (!isMarkdown(html)) {
    return service.turndown(html).trim();
  }
  return html.trim();
}
