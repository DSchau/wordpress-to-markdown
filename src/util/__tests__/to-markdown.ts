import { isMarkdown, toMarkdown, addjsToCodeFence } from '../to-markdown';

test('it detects markdown', () => {
  [
    `# Hello World`,
    '```\nalert("hello world")```',
    `
some content
___
    `,
  ].forEach(sample => {
    expect(isMarkdown(sample)).toBe(true);
  });
});

test('it translates html to markdown', async () => {
  expect(
    await toMarkdown(`
<h1>Hello World</h1>

<ul>
  <li>Item One</li>
  <li>Item Two</li>
</ul>

<hr>
  `)
  ).toBe(
    `
# Hello World

-   Item One
-   Item Two

___
  `.trim()
  );
});

describe('pre translation to code fence', () => {
  test('it translates simple pre block', async () => {
    expect(
      await toMarkdown(`
<pre>
alert('hello world');

alert('hi');
</pre>
    `)
    ).toBe(
      `
\`\`\`
alert('hello world');

alert('hi');
\`\`\`
    `.trim()
    );
  });

  test('it translates pre block with language', async () => {
    expect(
      await toMarkdown(`
<pre lang="groovy">
import groovy.xml.MarkupBuilder

def xml = new MarkupBuilder()
</pre>
    `)
    ).toBe(
      `
\`\`\`groovy
import groovy.xml.MarkupBuilder

def xml = new MarkupBuilder()
\`\`\`
    `.trim()
    );
  });

  test('it does _not_ apply code fence to a block without a new line', async () => {
    expect(
      await toMarkdown(`
<pre lang="js">
var a = 'b';
</pre>
    `)
    ).toBe("`var a = 'b';`");
  });

  test('addjs', async () => {
    const result = await addjsToCodeFence(
      '[addjs src="https://gist.github.com/JakePartusch/50737c37e988759e66b6.js?file=angular-controller.js"\\]'
    );
    expect(result).toMatchSnapshot();
  });

  test('it translates addjs links', async () => {
    expect(
      await toMarkdown(`
    Angular 1.x: \[addjs src="https://gist.github.com/JakePartusch/50737c37e988759e66b6.js?file=angular-controller.js"\] Angular 2.0: \[addjs src="https://gist.github.com/JakePartusch/d5863172113a92fc493d.js?file=angular2-component.ts"\]
    `)
    ).toMatchSnapshot();
  });
});
