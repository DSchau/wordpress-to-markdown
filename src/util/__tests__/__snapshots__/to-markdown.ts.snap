// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pre translation to code fence addjs 1`] = `
"

\`\`\`js
angular.module(‘example’)
	.controller(‘ExampleCtrl’, function() {
});
\`\`\`

"
`;

exports[`pre translation to code fence it translates addjs links 1`] = `
"Angular 1.x: \\\\\`\`\`
    angular.module(‘example’)
	.controller(‘ExampleCtrl’, function() {
});
  \`\`\` Angular 2.0: \\\\\`\`\`
    import {Component, View} from 'angular2/angular2';

@Component({
  selector: 'example'
})

@View({
  templateUrl: './components/example/example.html'
})

export class Example {}
  \`\`\`"
`;

exports[`pre translation to code fence it translates pre block with language 1`] = `
"\`\`\`groovy
import groovy_xml_MarkupBuilder
def xml = new MarkupBuilder
\`\`\`
"
`;
